src/haptic_feedback.c

```diff
@@ -0,0 +1,183 @@
+    haptic_feedback->alt_timer = current_time;
+    haptic_feedback->haptic_feedback_state.type = HAPTIC_FEEDBACK_NONE;
+    haptic_feedback->counter = 0;
+    haptic_feedback->duty.peroid = 200;  // 200 ms
+    haptic_feedback->error.peroid = 200;  // 200 ms
+    haptic_feedback->note_duration = 0;
+    haptic_feedback->beep_num = 0;
+}
+
+void haptic_feedback_play_single_tone(float freq, float volt, int channel) {
+    if (!VESC_IF->foc_play_tone) {
+        return;
+    }
+    if ((volt > 0) && (freq > 0.0)) {
+        VESC_IF->foc_play_tone(channel, freq, volt);
+    }
+}
+
+void haptic_feedback_play_two_tones(HapticFeedback *haptic_feedback) {
+    haptic_feedback_play_single_tone(
+        haptic_feedback->haptic_feedback_state.freq, haptic_feedback->haptic_feedback_state.volt, 0
+    );
+    haptic_feedback_play_single_tone(
+        haptic_feedback->vibrate_freq, haptic_feedback->vibrate_volt, 1
+    );
+}
+
+void haptic_feedback_check_tiltback(
+    HapticFeedback *haptic_feedback, State *state, float current_time, float duty_cycle
+) {
+    if (!(VESC_IF->foc_play_tone && VESC_IF->foc_stop_audio)) {
+        return;
+    }
+
+    if (state->mode == MODE_FLYWHEEL) {
+        haptic_feedback_reset(haptic_feedback, current_time);
+        return;
+    }
+
+    if (state->sat >= SAT_PB_DUTY) {
+        HAPTIC_FEEDBACK_TYPE prev = haptic_feedback->haptic_feedback_state.type;
+        if (state->sat == SAT_PB_DUTY) {
+            haptic_feedback->haptic_feedback_state = haptic_feedback->duty;
+            if ((duty_cycle >= haptic_feedback->duty_solid)) {
+                if (haptic_feedback->counter % 2 == 0) {
+                    haptic_feedback->alt_timer = current_time;
+                }
+            }
+        } else {
+            haptic_feedback->haptic_feedback_state = haptic_feedback->error;
+            switch (state->sat) {
+            case SAT_PB_TEMPERATURE:
+                // Careful
+                if (haptic_feedback->haptic_feedback_state.peroid > 0) {
+                    haptic_feedback->haptic_feedback_state.peroid /= 2;
+                }
+                __attribute__((fallthrough));
+            case SAT_PB_LOW_VOLTAGE:
+                __attribute__((fallthrough));
+            case SAT_PB_HIGH_VOLTAGE:
+                if (haptic_feedback->counter == 3) {
+                    haptic_feedback->haptic_feedback_state.peroid *= 2;
+                } else if (haptic_feedback->counter == 4) {
+                    haptic_feedback->counter = 0;
+                }
+                break;
+            default:
+                break;
+            }
+        }
+        // Reset state if a different type of event is coming in
+        if (haptic_feedback->haptic_feedback_state.type != prev) {
+            haptic_feedback_reset(haptic_feedback, current_time);
+        }
+
+        if (!haptic_feedback->tone_in_progress) {
+            haptic_feedback_play_two_tones(haptic_feedback);
```

Comment:
In a clean design this call shouldn't be here. This Function should just set up the tones or the pattern, and the `haptic_feedback_update()` function take care of the playing by itself. I understand the logic you have relies on this, see my comment on the `haptic_feedback_update()` function.

Reply:
If you don't immediately start playing a tone when called, then by the time you get to the haptic_feedback_update function timer and alt_timer will already be out of date.